FirebaseFunctions/Sources/Internal/FunctionsSerializer.swift

```diff
@@ -20,21 +20,21 @@ private enum Constants {
   static let dateType = "type.googleapis.com/google.protobuf.Timestamp"
 }
 
-enum SerializerError: Error {
-  // TODO: Add parameters class name and value
-  case unsupportedType // (className: String, value: AnyObject)
-  case unknownNumberType(charValue: String, number: NSNumber)
-  case invalidValueForType(value: String, requestedType: String)
+extension FUNSerializer {
+  enum Error: Swift.Error {
```

Comment:
This LGTM! I wanted to leave a rationale as to why I think why in case this serves as precedent for the future. I don't think the repo has a standard way when it comes to defining errors within vs. alongside the type it's associated with (neither do the Swift docs?). In this case, I think it makes sense to encapsulate the error in the type because the error is tightly coupled to only this class (e.g. only serializer throws serializer errors). There is one code path I'm seeing where this error type bubbles up out of a public API. Since the error type isn't public (maybe it should be?), users can't explicitly catch it. If the error needed to be public, the encapsulating type would need to be made public. I think `FunctionsSerializer.Error` reads well, but it may be more consistent to have it separate (`SerializerError`). For now, I think this is a good improvement and probably a better direction to go for defining specific Swift errors 👍

Reply:
I’d say it was pure luck (or coincidence) this time: an internal component, no protocols, no `@objc`, etc. 😅

I agree with your rationale, except the point on making the error public: If the user doesn’t see the internal component itself, I don’t think they should see or try and catch its specific errors. For clarity, the _public_ component could rethrow the error without exposing its type (something like `FunctionsError.serializationError(any Error)`). Plus there’s always `LocalizedError` for providing meaningful explanations instead of `SomeInternalError Error 0`.

Reply:
> For clarity, the _public_ component could rethrow the error without exposing its type (something like `FunctionsError.serializationError(any Error)`). Plus there’s always `LocalizedError` for providing meaningful explanations instead of `SomeInternalError Error 0`.

Ah, I see, thanks! This sounds like a good path forward if/when needed.